plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,278 @@
+lookupHeapLoc :: Name () -> Environment -> HeapLoc
+lookupHeapLoc name (Environment env) =
+    case IntMap.lookup (unUnique $ nameUnique name) env of
+      Nothing  -> throw $ LMachineStringException ("Name " ++ show (nameString name) ++ " missing from environment")
+      Just loc -> loc
+
+emptyHeap :: Heap
+emptyHeap = Heap IntMap.empty 0
+
+-- | Insert a new entry into a heap. The heap grows monotonically with no garbage collection or reuse of heap slots.
+insertInHeap :: Closure -> Heap -> (HeapLoc, Heap)
+insertInHeap cl (Heap h top) =
+    let top' = top + 1
+    in (top', Heap (IntMap.insert top' (Unevaluated cl) h) top')
+
+-- | Replace the heap entry at a given location.
+-- The old entry presumably persits in the old heap.
+updateHeap :: HeapLoc -> HeapEntry -> Heap -> Heap
+updateHeap l cl (Heap h top) =
+    Heap (IntMap.insert l cl h) top
+    -- insert seems to be faster than adjust (18s vs 28 s for fib 32) and also uses less memory.
+
+lookupHeap :: HeapLoc -> Heap -> HeapEntry
+lookupHeap l (Heap h _) =
+    case IntMap.lookup l h of
+      Just e  -> e
+      Nothing -> throw $ LMachineStringException ("Missing heap location in lookupHeap: " ++ show l)  -- This should never happen
+
+
+-- | The basic computation step of the L machine.  Search down the AST looking for a value, saving surrounding contexts on the stack.
+computeL :: EvaluationContext -> Heap -> Closure -> LMachineResult
+computeL ctx heap cl@(Closure term env) =
+    case term of
+      TyInst _ fun ty       -> computeL (FrameTyInstArg ty : ctx)                 heap (Closure fun env)
+      Apply _ fun arg       -> computeL (FrameAppArg (Closure arg env) : ctx)     heap (Closure fun env)
+      Wrap ann tyn ty term' -> computeL (FrameWrap ann tyn ty : ctx)              heap (Closure term' env)
+      Unwrap _ term'        -> computeL (FrameUnwrap : ctx)                       heap (Closure term' env)
+      TyAbs{}               -> returnL  ctx heap cl
+      LamAbs{}              -> returnL  ctx heap cl
+      Builtin{}             -> returnL  ctx heap cl
+      Constant{}            -> returnL  ctx heap cl
+      Error{}               -> Failure
+      Var _ name            -> let l = lookupHeapLoc name env
+                               in case lookupHeap l heap of
+                                    Evaluated cl'   -> returnL ctx heap cl'
+                                    Unevaluated cl' -> computeL (FrameHeapUpdate l : ctx) heap cl'
+
+
+-- | Return a closure containing a value. Ideally the fact that we've
+-- got a value would be enforced in the (Haskell) type. Values still
+-- have to be contained in closures.  We could have something like
+-- @v= \x.y@, where @y@ is bound in an environment.  If we ever go on
+-- to apply @v@, we'll need the value of @y@.
+returnL :: EvaluationContext -> Heap -> Closure -> LMachineResult
+returnL [] heap res                                                        = Success res heap
```

Comment:
Why not use a case on the continuation here?

Reply:
Good question. I'm not sure why I did that.  It'd probably make the code a lot narrower here, which would be good.

Reply:
Done.